### PR TITLE
fix: incorrect column pruning in sql with window operations

### LINKED ISSUE #6001: Incorrect column pruning in sql with window operations
### Describe the bug

As the title

### To Reproduce

```sql
❯ explain  select sum(case when latitude < 50.0 then latitude else 0 end) over (partition by name) from readings;
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Int64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |   WindowAggr: windowExpr=[[SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Float64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING AS SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Int64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING]]                                                                                                                                                                                                                 |
|               |     TableScan: readings projection=[time, name, fleet, driver, model, device_version, latitude, longitude, elevation, velocity, heading, grade, fuel_consumption, load_capacity, fuel_capacity, nominal_fuel_consumption]                                                                                                                                                                                                                                                                                                                                                                                     |
| physical_plan | ProjectionExec: expr=[SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Int64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING@16 as SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Int64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING]                                                                                                                                                                                                                      |
|               |   WindowAggExec: wdw=[SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Int64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING: Ok(Field { name: "SUM(CASE WHEN readings.latitude < Float64(50) THEN readings.latitude ELSE Int64(0) END) PARTITION BY [readings.name] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(NULL)), end_bound: Following(UInt64(NULL)) }] |
|               |     SortExec: expr=[name@1 ASC NULLS LAST]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |       CoalesceBatchesExec: target_batch_size=8192                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
|               |         RepartitionExec: partitioning=Hash([Column { name: "name", index: 1 }], 8), input_partitions=1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
|               |           ParquetExec: limit=None, partitions={1 group: [[Users/yukkit/Documents/tmp/data/parquet/part-297.parquet]]}, projection=[time, name, fleet, driver, model, device_version, latitude, longitude, elevation, velocity, heading, grade, fuel_consumption, load_capacity, fuel_capacity, nominal_fuel_consumption]                                                                                                                                                                                                                                                                                      |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set. Query took 0.026 seconds.
```

### Expected behavior

Push down only `latitude` and `name`  to TableScan.

### Additional context

version commit id: bd705fe6

### PR DESCRIPTION
**Pull request by yukkit**
# Which issue does this PR close?

Closes #6001 .

# Rationale for this change

# What changes are included in this PR?

If the node that window's child has outputs not in referenced inputs, then add a projection

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/push_down_projection.rs [hunk 1]
```diff
@@ -312,11 +312,31 @@ impl OptimizerRule for PushDownProjection {
 
                 if new_window_expr.is_empty() {
                     // none columns in window expr are needed, remove the window expr
-                    let new_window = window.input.as_ref().clone();
+                    let new_window = window.input.clone();
+
+                    let new_window = restrict_outputs(&new_window, &required_columns)?
+                        .unwrap_or((*new_window).clone());
```
Comment (jackwener) [inline datafusion/optimizer/src/push_down_projection.rs:318]:
```suggestion
                    let input = window.input.as_ref();

                    let new_input = restrict_outputs(input, &required_columns)?
                        .unwrap_or(input.clone());
```

File: datafusion/optimizer/src/push_down_projection.rs [hunk 2]
```diff
@@ -312,11 +312,31 @@ impl OptimizerRule for PushDownProjection {
...
+                        .for_each(|f| {
+                            referenced_inputs.insert(f.qualified_column());
+                        });
+
+                    let window_child = window.input.clone();
+                    let new_window_child =
+                        restrict_outputs(&window_child, &referenced_inputs)?
+                            .unwrap_or((*window_child).clone());
+
+                    let new_window = LogicalPlanBuilder::from(new_window_child)
```
Comment (jackwener) [inline datafusion/optimizer/src/push_down_projection.rs:339]:
```suggestion
                    let input = window.input.as_ref();
                    let new_input = restrict_outputs(input, &referenced_inputs)?
                        .unwrap_or(input.clone());

                    let new_window = LogicalPlanBuilder::from(new_input)
```

File: datafusion/optimizer/src/push_down_projection.rs [hunk 3]
```diff
@@ -553,6 +573,21 @@ fn push_down_scan(
...
+    permitted_outputs: &HashSet<Column>,
+) -> Result<Option<LogicalPlan>> {
+    let schema = plan.schema();
+    if permitted_outputs.len() == schema.fields().len() {
+        return Ok(None);
+    }
+    Ok(Some(generate_projection(
+        permitted_outputs,
+        schema,
+        plan.clone(),
```
Comment (jackwener) [inline datafusion/optimizer/src/push_down_projection.rs:587]:
```suggestion
        Arc::new(plan.clone()),
```

**Review by jackwener — APPROVED** [review]
A great job, Thank you @yukkit

**Comment by yukkit** [comment]
@jackwener Thanks for your suggestion. I modified the variable naming, but I don't understand the modification here   https://github.com/apache/arrow-datafusion/pull/6039#discussion_r1170046496 as it results in a clone LogicalPlan instead of a pointer, I expect to know your thoughts.

**Comment by jackwener** [comment]
> @jackwener Thanks for your suggestion. I modified the variable naming, but I don't understand the modification here [#6039 (comment)](https://github.com/apache/arrow-datafusion/pull/6039#discussion_r1170046496) as it results in a clone LogicalPlan instead of a pointer, I expect to know your thoughts.

Oh, It's my fault, I don't notice it. Thank you. @yukkit

**Pull request merged**